docs/howto/mocking_services.md

```diff
@@ -0,0 +1,273 @@
+import zio.macros.access.Accessable
+import zio.macros.mock.Mockable
+import zio.console.Console
+
+@Accessable
+@Mockable
+trait ProductObserver {
+  val productObserver: ProductObserver.Service[Any]
+}
+
+object ProductObserver {
+  trait Service[R] {
+    def processEvent(event: ProductEvent): ZIO[R, Nothing, Unit]
+  }
+
+  // autogenerated `object Service { ... }`
+  // autogenerated `object > extends Service[ProductObserver] { ... }`
+  // autogenerated `implicit val mockable: Mockable[ProductObserver] = ...`
+
+  trait Live extends ProductObserver {
+    // dependency on Console module
+    val console: Console.Service[Any]
+
+    val productObserver = new Service[Any] {
+      def processEvent(event: ProductEvent): UIO[Unit] =
+        for {
+         _    <- console.putStrLn(s"Got $event")
+         line <- console.getStrLn.orDie
+         _    <- console.putStrLn(s"You entered: $line")
+        } yield ()
+    }
+  }
+}
+```
+
+## Setting up expectations
+
+Finally we're all set and can create ad-hoc mock environments with our services.
+
+```scala mdoc:silent
+import zio.test._
+import zio.test.Assertion._
+
+val event = new ProductEvent
+val app = ProductObserver.>.processEvent(event)
+val mockEnv: Managed[Nothing, MockConsole] = (
+  MockSpec.expectIn(MockConsole.Service.putStrLn)(equalTo(s"Got $event")) *>
+  MockSpec.expectOut(MockConsole.Service.getStrLn)("42") *>
+  MockSpec.expectIn(MockConsole.Service.putStrLn)(equalTo("You entered: 42"))
+)
+```
+
+## Providing mocked environment
+
+```scala mdoc:silent
+object ProductObserverSpec extends DefaultRunnableSpec(
+  suite("processEvent")(
+    testM("calls putStrLn > getStrLn > putStrLn and returns unit") {
+      val result = app.provideManaged(mockEnv)
+      assertM(result, isUnit)
+    }
+  )
+)
+```
+
+## Mocking multiple collaborators
+
+In some cases we have more than one collaborating service being called. In such situations we need to build our expectations seperately for each
+service and then combine them into single environment:
+
+```scala mdoc:silent
+val mockConsole: Managed[Nothing, MockConsole] = (
+  MockSpec.expectIn(MockConsole.Service.putStrLn)(equalTo("What is your name?")) *>
+  MockSpec.expectOut(MockConsole.Service.getStrLn)("Mike")) *>
+  MockSpec.expectIn(MockConsole.Service.putStrLn)("Mike, your lucky number today is 42!"))
+)
+
+val mockRandom: Managed[Nothing, MockRandom] =
+  MockSpec.expectOut(MockRandom.Service.nextInt._1)(42))
+
+val combinedEnv =
+  (mockConsole &&& mockRandom)
+    .map {
+      case (c, r) => new Console with Random {
+       val console = c.console
+       val random = r.random
+      }
+    }
+
+val combinedApp =
+  for {
+    _    <- console.putStrLn("What is your name?")
+    name <- console.getStrLn.orDie
+    _    <- console.putStrLn(s"$name, your lucky number today is 42!")
+  } yield ()
+
+val result = combinedApp.provideManaged(combinedEnv)
+assertM(result, isUnit)
+```
+
+[doc-module-pattern](module_pattern.md)
```

Comment:
Isn't the correct syntax `[doc-module-pattern]: module_pattern.md` ?

Reply:
Yes :+1: